002-per-isolate-exit-hooks.md

```diff
@@ -0,0 +1,54 @@
+| Title  | Per-isolate exit hooks      |
+|--------|-----------------------------|
+| Author | @dherman                    |
+| Status | DRAFT                       |
+| Date   | 2016-02-12                  |
+
+## Description
+
+The public C++ API has the [`AtExit`](https://nodejs.org/api/addons.html#addons_void_atexit_callback_args) API for specifying teardown logic for a Node process, but with the upcoming addition of workers, there will be a need for custom teardown logic for arbitrary isolates.
+
+This is particularly useful for addons that want to lazily allocate per-isolate data on a by-need basis, and therefore do not control the actual creation of the workers. Addons like that need the ability to deallocate any associated data structures when the isolate goes away.
+
+For example, an addon can create a tagged native class with [`v8::FunctionTemplate`](https://v8docs.nodesource.com/node-5.0/d8/d83/classv8_1_1_function_template.html) that's associated with some additional class metadata. (This is in fact my situation for [https://github.com/rustbridge/neon](Neon).) Any workers that use the class will cause it to lazily create the class, its function template, and its metadata. In order to avoid a memory leak, the code that lazily allocates the metadata will also want to register a callback to deallocate that metadata when the worker terminates.
+
+## API
+
+**enum IsolateKind { Main, Worker }**
+  * **`Main`**: this isolate belongs to the main Node process.
+  * **`Worker`**: this isolate belongs to a worker.
+
+**void AtIsolateExit(isolate, callback, args)**
```

Comment:
@bnoordhuis how would that look? just move `isolate` to the end and put `void*` at the start of the callback signature? +1 if we can achieve overloading here.